### ISSUE TITLE
blits/bleps/blamps, polynomial and table-based, linear- and minimum-phase

### ISSUE THREAD

**Issue by RobinSchmidt** [RobinSchmidt/RS-MET#273]
i took the few offline days as opportunity to work distraction free on an implementation of bandlimited impulses, steps and ramps. it may seem seem weird that i don't already have this kind of stuff in the codebase, but it just recently revealed itself to me, how to implement the overlap efficiently and elegantly: you don't loop over the active bleps per sample, but pre-render the whole blep into a circular buffer (and accumulate more of them into the same buffer, if they happen to overlap) and per sample, juts read out the buffer content per sample, add it and clear it. works like a charm. here's a square-wave, anti-aliased with table-based lin- and min-blep

![image](https://user-images.githubusercontent.com/1833099/56043336-557aa500-5d3d-11e9-981f-a4464fd4448e.png)

i have poly-bleps, too. ...and blits and blamps. this can be applied to anti-alias any signal generator algorithm that produces step- and corner discontinuitities (impulses too - but that's probably not needed)......blamps for corners....that makes it ideally suited to apply to turtle-graphics (snowflake) but also the trisaw osc.....

**Comment by RobinSchmidt** [comment]
they are still in the rs_testing module, but i think can soon be transferred to the library proper...a few things still have to be sorted out and cleaned up, though

**Comment by RobinSchmidt** [comment]
the table-based linbleps are based on an elliptic filter - the idea could be extended to any of EngineersFilter's options - we may later have butterbleps, besselbleps, chebybleps - whatever

**Comment by elanhickler** [comment]
GOOOOOOOOOOOD!!!!!!!!!!!!!!!!!!!!!

Does this also antialias the corners of a triangle wave?

**Comment by RobinSchmidt** [comment]
yes! the blamp does!

**Comment by elanhickler** [comment]
Ok, how about discontinuities in my oscilloscope music synths? Some of them have harsh resets or jumps, but it depends on the settings. It should be easy to predict with most settings.

**Comment by RobinSchmidt** [comment]
yes - as siad - any sort of generator that produces steps and/or corners should be the perfect target for this stuff. the generator just needs to supply the sub-sample position of the discontinuity and its size. the rest is then taken care of by my blep/blamp object

**Comment by elanhickler** [comment]
YESSSSSSSSSSSSSSSSSSSSSSSSsssssssssssssss. Let me know when you implement officially in TriSaw so I can update my synths.

**Comment by RobinSchmidt** [comment]
oh - that sounds good! the main problem i see is that those fractal graphics may produce a fucking lot of corners, so applying a blamp to each of them might become computationally expensive - but in principle, it should be straightforward. i also added to the original post, that this should also be applicable to raybouncer. i was dragged into the whole blit/blep/blamp rabbithole by this thread:

https://www.kvraudio.com/forum/viewtopic.php?f=33&t=522701

and after reading these papers:

https://quod.lib.umich.edu/cgi/p/pod/dod-idx/continuous-order-polygonalwaveform-synthesis.pdf?c=icmc;idno=bbp2372.2016.104;format=pdf
http://www.dafx17.eca.ed.ac.uk/papers/DAFx17_paper_100.pdf

this is all very interesting stuff

**Comment by elanhickler** [comment]
oh of course, xoxos.

**Comment by elanhickler** [comment]
wait a minute wait a minute............... what about modulation? The edges can be impossible to predict. How many samples of delay do you need?

**Comment by RobinSchmidt** [comment]
hmm...the delay depends on the particular flavor of the blep. the table-based linear-phase ones introduce a delay of half of the kernel length, the polybleps only one or two samples (which is also half-the-kernel-length because they correct only 2 or 4 samples directly around the discontinuity - which may be already surprisingly effective). for the minimum-phase table-based blep (aka "minblep"), it's a bit more complicated, because the delay is - strictly speaking - frequency-dependent. loosely-speaking and just eyeballing it, it's about 3 samples though (for the currently implemented elliptic filter based blep).

but this delay does not really mean that you need to be able to predict the discontinuity that many samples beforehand. if your trivial osc can say: in between the next sample (i.e. the next call to blep.getSample) and the previous one, a step occurred (at subsample-position of blah and with an amplitude of blub), this should be enough.

i think, it would be most convenient to have a sort of driver object wrapped around the osc and blep(per) and then do something like:

```cpp
MyOscDriver::getSample()
{
   double x = osc.getSample(); // osc produces naive sample for this instant
   
   if(osc.getStepAmplitude() != 0.0) // a step did occur between this and the previous sample
     blep.prepareForStep(osc.getStepDelay(), osc.getStepAmplitude());  
     // tell the blep the sub-sample instant of the step and its height - the osc must be able to
     // supply this information
     
   return blep.getSample(x); // return sample with blep-correction applied
}
```

that "driver" object could also maintain a pair or bunch of oscs and manage blepped hardsync. i think, that's a very flexible way of implementing the blep correction. the implementations, i've seen before (on kvr and elsewhere) always had the signal-generation and blep-anti-aliasing intertwined in a monolithic class. my implementation is modular and can be *applied* any existing naive generator as post-processing step (provided that this naive generator supplies the required information).
